File: code/examples/ts/06_webhooks.ts
/**
 * 06 — Webhooks
 *
 * Register a server-side webhook so Sifter POSTs to your endpoint
 * whenever a document is processed or a sift completes.
 *
 * This example registers the hook, uploads documents, then
 * simulates receiving events via a minimal HTTP server.
 *
 * Requirements:
 *   npx tsx 06_webhooks.ts
 *   # Sifter server running on localhost:8000
 *   # Your webhook endpoint must be reachable from the Sifter server
 *   # SIFTER_API_KEY env var set
 */
import http from "http";
import { Sifter } from "@sifter-ai/sdk";

// ── Minimal webhook receiver ──────────────────────────────────────────────────

const receivedEvents: Array<Record<string, unknown>> = [];

const server = http.createServer((req, res) => {
  if (req.method !== "POST") { res.end(); return; }
  const chunks: Buffer[] = [];
  req.on("data", (c: Buffer) => chunks.push(c));
  req.on("end", () => {
    try {
      const event = JSON.parse(Buffer.concat(chunks).toString()) as Record<string, unknown>;
      receivedEvents.push(event);
      const payload = (event["payload"] as Record<string, unknown> | undefined) ?? {};
      const docId = (payload["document_id"] as string) ?? "—";
      console.log(`\n[webhook] event=${event["event"]} doc=${docId}`);
    } catch { /* ignore malformed */ }
    res.writeHead(200).end();
  });
});

await new Promise<void>(resolve => server.listen(9000, resolve));
console.log("Webhook receiver listening on http://localhost:9000");

// ── Register webhook with Sifter ──────────────────────────────────────────────

const s = new Sifter();

const hook = await s.registerHook(
  ["sift.document.processed", "sift.completed", "sift.error"],
  "http://localhost:9000",
);
console.log(`Webhook registered: ${(hook as { id: string }).id}`);

// ── Upload and process ────────────────────────────────────────────────────────

const sift = await s.createSift(
  "Webhook demo TS",
  "Extract the document type, date, parties involved and key amount.",
);

await sift.upload("../documents/invoices/", { onConflict: "replace" });
console.log("Processing... (webhook events will appear above)");
await sift.wait();

console.log(`\nTotal webhook events received: ${receivedEvents.length}`);
for (const ev of receivedEvents) {
  console.log(`  ${ev["event"]}`);
}

// Cleanup
await s.deleteHook((hook as { id: string }).id);
await sift.delete();
server.close();

File: code/examples/python/06_webhooks.py
"""
06 — Webhooks

Register a server-side webhook so Sifter POSTs to your endpoint
whenever a document is processed or a sift completes.

This example registers the hook, uploads documents, then
simulates receiving events via a minimal HTTP server.

Requirements:
    pip install sifter-ai httpx
    # Sifter server running on localhost:8000 (./run.sh)
    # Your webhook endpoint must be reachable from the Sifter server
"""
import json
import threading
from http.server import BaseHTTPRequestHandler, HTTPServer

from sifter import Sifter

# ── Minimal webhook receiver ──────────────────────────────────────────────────

received_events: list[dict] = []

class WebhookHandler(BaseHTTPRequestHandler):
    def do_POST(self):
        length = int(self.headers.get("Content-Length", 0))
        body = self.rfile.read(length)
        event = json.loads(body)
        received_events.append(event)
        payload = event.get('payload', {})
        doc_id = payload.get('document_id', '—')
        print(f"\n[webhook] event={event.get('event')} doc={doc_id}")
        self.send_response(200)
        self.end_headers()

    def log_message(self, *args):
        pass  # suppress access log

server = HTTPServer(("0.0.0.0", 9000), WebhookHandler)
thread = threading.Thread(target=server.serve_forever, daemon=True)
thread.start()
print("Webhook receiver listening on http://localhost:9000")

# ── Register webhook with Sifter ──────────────────────────────────────────────

s = Sifter()

hook = s.register_hook(
    events=["sift.document.processed", "sift.completed", "sift.error"],
    url="http://localhost:9000",  # must be reachable from the Sifter server
)
print(f"Webhook registered: {hook['id']}")

# ── Upload and process ────────────────────────────────────────────────────────

sift = s.create_sift(
    name="Webhook demo",
    instructions="Extract the document type, date, parties involved and key amount.",
)

sift.upload("../documents/invoices/", on_conflict="replace")
print("Processing... (webhook events will appear above)")
sift.wait()

print(f"\nTotal webhook events received: {len(received_events)}")
for ev in received_events:
    print(f"  {ev.get('event')}")

# Cleanup
s.delete_hook(hook["id"])
sift.delete()
server.shutdown()
